config/data_scrubber.go

```diff
@@ -0,0 +1,62 @@
+package config
+
+import (
+	"regexp"
+	"strings"
+)
+
+var (
+	defaultSenstiveWords = []string{"password", "passwd", "mysql_pwd", "access_token", "auth_token", "api_key", "apikey", "secret", "credentials", "stripetoken"}
+)
+
+type DataScrubber struct {
+	Enabled               bool
+	SensitiveWords        []*regexp.Regexp
+	DefaultSensitiveWords []*regexp.Regexp
+	CustomSensitiveWords  []*regexp.Regexp
+}
+
+func NewDefaultDataScrubber() *DataScrubber {
+	newDataScrubber := &DataScrubber{
+		Enabled:               true,
+		SensitiveWords:        CompileStringsToRegex(defaultSenstiveWords),
+		DefaultSensitiveWords: CompileStringsToRegex(defaultSenstiveWords),
+		CustomSensitiveWords:  make([]*regexp.Regexp, 0, 0),
+	}
+
+	return newDataScrubber
+}
+
+// Compile each word in the list into a regex pattern to match against the cmdline arguments
+func CompileStringsToRegex(words []string) []*regexp.Regexp {
+	compiledRegexps := make([]*regexp.Regexp, 0, len(words))
+	for _, word := range words {
+		pattern := `(?P<key>( |-)(?i)` + word + `)(?P<delimiter> +|=)(?P<value>[^\s]*)`
+		r := regexp.MustCompile(pattern)
+		compiledRegexps = append(compiledRegexps, r)
+	}
+
+	return compiledRegexps
+}
+
+// Hide any cmdline argument value whose key matchs one of the patterns on the argsBlacklist vector
+func (ds *DataScrubber) ScrubCmdline(cmdline []string) []string {
+	if !ds.Enabled {
+		return cmdline
+	}
+
+	rawCmdline := strings.Join(cmdline, " ")
+	for _, pattern := range ds.SensitiveWords {
+		rawCmdline = pattern.ReplaceAllString(rawCmdline, `${key}${delimiter}********`)
+	}
+
+	return strings.Split(rawCmdline, " ")
```

Comment:
Are we 100% certain that data will remain the same when we do the join/split and that there is no change anything of the args in `cmdline` will have a space?

Reply:
I can envision 2 cases where the cmdline can change:

1. It hasn't been splitted before when collected by AllProcesses(). So we'll end up splitting it. By printing the processes cmdline on "process.go", we verified that it happened sometimes.
2. It was partially splitted. If we have ["run", "-p 1234"] we'll produce ["run", "-p", "1234"]

I can prevent this modification for the cmdlines that don't match any pattern if it'd be better. At worst, we only modify the structure of the cmdlines that have an arg to be hidden. What do you think?

Reply:
Ok. As long as things work as expected on the backend in either case, it should be fine.